src/plugins/esql_datagrid/public/data_grid.tsx

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React, { useState, useCallback } from 'react';
+import { zipObject } from 'lodash';
+import { UnifiedDataTable, DataLoadingState } from '@kbn/unified-data-table';
+import { Storage } from '@kbn/kibana-utils-plugin/public';
+import type { ESQLRow } from '@kbn/es-types';
+import type {
+  DatatableColumn,
+  DatatableColumnMeta,
+  DatatableRow,
+} from '@kbn/expressions-plugin/common';
+import type { AggregateQuery } from '@kbn/es-query';
+import type { DataTableRecord } from '@kbn/discover-utils/types';
+import type { DataView } from '@kbn/data-views-plugin/common';
+import type { CoreStart } from '@kbn/core/public';
+import type { DataPublicPluginStart } from '@kbn/data-plugin/public';
+import type { FieldFormatsStart } from '@kbn/field-formats-plugin/public';
+import { RowViewer } from './row_viewer_lazy';
+
+interface ESQLDataGridProps {
+  core: CoreStart;
+  data: DataPublicPluginStart;
+  fieldFormats: FieldFormatsStart;
+  rows: ESQLRow[];
+  dataView: DataView;
+  columns: DatatableColumn[];
+  query: AggregateQuery;
+  flyoutType?: 'overlay' | 'push';
+  isTableView?: boolean;
+}
+type DataTableColumnsMeta = Record<
+  string,
+  {
+    type: DatatableColumnMeta['type'];
+    esType?: DatatableColumnMeta['esType'];
+  }
+>;
+
+const DataGrid: React.FC<ESQLDataGridProps> = (props) => {
+  const storage = new Storage(localStorage);
+  const [expandedDoc, setExpandedDoc] = useState<DataTableRecord | undefined>(undefined);
+  const [activeColumns, setActiveColumns] = useState<string[]>(
+    props.isTableView ? props.columns.map((c) => c.name) : []
+  );
+  const [rowHeight, setRowHeight] = useState<number>(5);
+
+  if (props.dataView.fields.getByName('@timestamp')?.type === 'date') {
+    props.dataView.timeFieldName = '@timestamp';
+  }
```

Comment:
Can we avoid `dataView` prop mutation?

Reply:
No why? We are doing this in multiple places, what are you afraid of?

Reply:
Mutating props can create unexpected behaviour in other components. Why can't we rather have this logic upper in the components tree, where the data view is being created?

Reply:
We are doing this everywhere we are creating an ES|QL adhoc dataview , the dataview is being created using the adhoc utility we are using across our apps. I would like to keep it as it is